api/wft-complete-metadata.md

```diff
@@ -0,0 +1,133 @@
+    { command_type: "local_activity", activity_name: "la_func", id: 2, retries: 5, location: "my_file.py:23" },
+    { command_type: "timer", id: 3, duration: "5s", location: "my_file.py:42" },
+]
+```
+
+### SDK internal versioning
+
+We would like to be able to make SDK changes which affect how history is written, which means
+flagging our behavior based on what version of the SDK the history was produced with. Note: *not*
+relating to concepts of versioning a *user's* code, but rather the SDK itself.
+
+This internal version could easily slot into this metadata.
+
+Note: This could potentially go in `temporal.api.taskqueue.v1.VersionId` too, but that's more
+intended to relate to user code.
+
+### Dynamic config adjustments
+
+Allow the worker to set configurable things which don't directly impact determinism or workflow
+state. For example, changing the WFT timeout for the next task.
+
+### Various debugging / analysis information
+
+There's a whole world of little bits of info that could be useful to enable selectively to help
+users debug their workflows, help us learn about performance information, etc.
+
+Some examples:
+* Emit the OTel span id that was associated with the WFT - now the user can go easily look up any
+  logs/events that were emitted as part of that WFT.
+* Metrics work great for all sorts of latency measurements, but what if you wanted to know about
+  how long a _specific_ WFT took? You could emit that.
+* Interceptors could pipe whatever arbitrary data the user wants into there, for use cases where
+  memo / search attributes / side effects may not be appropriate.
+
+### Things we haven't thought of
+
+Having this metadata be suitably generic means we dont need to update the API repo for every
+possible usecase. Particularly ones where the SDK is the only one who needs to be able to read it.
+
+## Implementation Proposal
+
+We'd add a new metadata field to the
+[WFT response](https://github.com/temporalio/api/blob/master/temporal/api/workflowservice/v1/request_response.proto#L274).
+And the [WFT completed event](https://github.com/temporalio/api/blob/f0350f8032ad2f0c60c539b3b61ea37f412f1cf7/temporal/api/history/v1/message.proto#L186).
+Though the completion event need not necessarily contain all the info sent in the completion. If
+it's known and understood that a particular metadata field is simply telling server something
+ephemeral, that could be stripped before writing the event.
+
+With the following proto as its type:
+
+```protobuf
+message StructuredMetadata {
+  // For JSON-esque data representations that can be easily decoded by anyone. Could contain user
+  // data that they'd like to show up the UI, for example.
+  map<string, google.protobuf.Struct> struct_metadata = 1;
+  // For better typed / more efficient binary serialization of types that are expected to be known
+  // by their consumers.
+  map<string, google.protobuf.Any> proto_metadata = 2;
+}
```

Comment:
```suggestion
message RespondWorkflowTaskCompletedRequest {
  // Other fields

  map<string, temporal.api.common.v1.Payload> metadata = 11;
}

message WorkflowTaskCompletedEventAttributes {
  // Other fields

  map<string, temporal.api.common.v1.Payload> metadata = 6;
}
```

I think this has over-thought the issue. I think we can just use normal payloads and have reserved field names for ourselves. Our UI knows how to show user-structured payloads already and our codecs understand that payloads should be encrypted and users may want to encrypt this information.

If we wanted to keep the `StructuredMetadata` wrapper message, that's fine too. But I think we need to reuse our existing structured metadata mechanisms for structured metadata. We can discuss if certain keys should use the default converters (i.e. unencrypted).

Reply:
I actually like the use of structured data here assuming that this data doesn't need to be encrypted.

Reply:
Having both the Any and the Struct maps isn't great - it's hard to predict or document where a given field will be as semantically they're the same. But we're really hamstrung by the protobuf type system here so the options aren't great. Do we need the two layers of lookup provided by a map of structs? Since a struct is itself a map it feels like we could get away with just a single Struct instance. What do you think about a convention where a `meta` field is always a Struct and if the Value of one of the Struct fields has to be a Payload in the future then so be it.

Reply:
I'd be fine with just one `Struct` which can potentially have `Payload` (or `Any`) values in it.

Reply:
I still think payload works very well for structured data and the UI knows how to deal with structured payload data and you get the bonus of encryptability and easy user value decoding. There are so many wins to using the same thing we use for memo, failure attributes, etc.

Reply:
It works... OK. Something like `Struct` internally probably would've made more sense in them as our default for things we want renderable in the UI rather than serialized JSON which is a bit wasteful, but that's in the past so w/e. 

I take your point about just being able to re-use it.

Honestly, none of this is terribly consequential. They're all equivalent and it's just a matter of what's easier in what situation. I think we should just pick something and move on. I like `Struct` because it sets us up to easily and efficiently represent what I think will be the case with all or the vast majority of the data represented here, which is data that won't ever be passed through a custom codec. If we'd rather just keep using Payload b/c it exists, I get that too. @bergundy if you can tie break, I'd say let's do that and move forward (or maybe we can grab someone from server team)

Reply:
The main question is if we want users to be able to put their (potentially sensitive) data in these fields.
If that's not a requirement, we could go with Structs which are easier for us to work with and don't require extra serialization. If we're unsure, I'd go with @cretz's suggestion of using `Payload`s.

Reply:
It's not just sensitive data. You're missing the other benefits I mentioned. What about the ability to easily decode to language-defined types in all languages? What about the consistency with all other metadata forms we use?

Reply:
I'm convinced.

Reply:
OK, let's go with Payloads then.

I just want to say though that
> There just isn't a good reason for untyped structs

This isn't true - there is a pretty good reason, which is serializing and encoding JSON strings is wasteful by comparison, and they are easier to work with for clients who don't have access to all our payload machinery. Now, I'm fine with saying that's not worth re-using all the payloads stuff for this case - but that doesn't mean there's literally no merit to the idea.

Reply:
> when we have a format that is perfect for this

Agree there is value in structs and other WKT for proto. For example Envoy uses proto to define their config formats and accepts these types for unknown user input instead of JSON.

IME though, working with these types in code makes you long for the typed protos or language structs you use otherwise. I've seen people just convert these things to JSON just so they can convert them into their language-typed forms anyways. But yes, they do have generic value.